comparison/comparison.md

```diff
@@ -0,0 +1,523 @@
+# Comparison with PyTorch's DataLoader
+
+```python
+import lazy_dataset
+from torch.utils import data as torch_data
+from torch.utils.data import DataLoader as TorchDataLoader
+```
+
+| Example | Feature | `lazy_dataset` | `TorchDataLoader` |
+| ------- | ------- | :------------: | :---------------: |
+|    1    | Dataset class | `lazy_datsaset.Dataset`| `torch_data.Dataset` |
+|    2    | Batching | yes | yes |
+|    2    | Collate batch | manual | automatic |
+|    3    | Shuffle | yes | yes |
+|    4    | Multi-process data loading | yes | yes (built-in shared memory GPU) |
+|    4    | Multi-thread data loading | yes (default) | no |
+|    see [README](../README.md)    | Filter (builtins.filter) | yes | no |
+|    5    | Sort by key | yes | no |
+|    5    | Sort by value | yes | no |
+|    6    | Draw random example | yes | no |
+|    7    | Unbatch | yes | no |
+
+## Examples
+1. Dataset
+    ```python
+    >>> examples = {
+    ...    'ex_1': {
+    ...        'example_id': 'ex_1',
+    ...        'observation': [1, 2, 3],
+    ...        'label': 1
+    ...    },
+    ...    'ex_2': {
+    ...        'example_id': 'ex_2',
+    ...        'observation': [4, 5, 6],
+    ...        'label': 2
+    ...    }
+    ... }
+
+    >>> class TorchDictDataset(torch_data.Dataset):
+    ...
+    ...    def __init__(self, examples):
+    ...        self.examples = examples
+    ...
+    ...    def __len__(self):
+    ...        return len(self.examples)
+    ...
+    ...    def __getitem__(self, idx):
+    ...        idx_to_key = list(self.examples.keys())[idx]
+    ...        return self.examples[idx_to_key]
+
+    >>> torch_dict_ds = TorchDictDataset(examples)
+    >>> print(torch_dict_ds[0])
+    {'example_id': 'ex_1', 'observation': [1, 2, 3], 'label': 1}
+
+    >>> dict_ds = lazy_dataset.from_dict(examples)
+    >>> print(dict_ds[0])
+    {'example_id': 'ex_1', 'observation': [1, 2, 3], 'label': 1}
+    ```
+2. Batching & Collate
+    ```python
+    >>> data_loader = TorchDataLoader(
+    ...     torch_dict_ds, batch_size=2, collate_fn=None
+    ... )
+    >>> next(iter(data_loader))
+    {'example_id': ['ex_1', 'ex_2'],
+     'observation': [tensor([1, 4]), tensor([2, 5]), tensor([3, 6])],
+     'label': tensor([1, 2])}
```

Comment:
Can you change the `observation` type to a `numpy.array` for the `torch.utils.data.DataLoader`? This output is strange and using list with different lengths the last element gets dropped:
```
In [25]: examples = { 
    ...:  
    ...:     'ex_1': { 
    ...:         'example_id': 'ex_1', 
    ...:         'observation': [1, 2, 3], 
    ...:         'label': 1 
    ...:     }, 
    ...:     'ex_2': { 
    ...:         'example_id': 'ex_2', 
    ...:         'observation': [5, 6], 
    ...:         'label': 2 
    ...:     } 
    ...:  }
In [26]: torch_dict_ds = TorchDictDataset(examples)                             
In [27]: data_loader = TorchDataLoader(torch_dict_ds, batch_size=2, collate_fn=None)                                                                   
In [28]: next(iter(data_loader))                                                
Out[28]: 
{'example_id': ['ex_1', 'ex_2'],
 'observation': [tensor([1, 5]), tensor([2, 6])],
 'label': tensor([1, 2])}

```

Reply:
Indeed, that is a weird behavior. I didn't notice it previously. Fixed in https://github.com/fgnt/lazy_dataset/pull/27/commits/33684af06b09d9c28befb42d549b2f7027585261